File: docs/languages/racket/lectures/primes.go
// primes.go

//
// Write a function alled is_prime(n) that returns true if n (and int) is a
// prime number, and false otherwise.
//
// Also write a function called primes_less_than(n) that returns the count of
// the number of integers less than n that are prime.
//
// An integer is prime if it is gSreater than 1, and has exactly two
// divisors, 1 and itself. For example, 2, 3, 5, and 101 are all prime,
// but 9, 15, 77, and 1001 are not prime.
//

package main

import "fmt"

func main() {
    for i := 0; i <= 25; i++ {
        fmt.Println(i, is_prime(i))
    }

    fmt.Println(primes_less_than(10000)) // correct answer: 1229
}

// Returns true if the integer n is prime, and false otherwise.
func is_prime(n int) bool {
    if n < 2 {
        return false
    } else if n == 2 {
        return true
    } else if n % 2 == 0 {
        return false
    } else {
        candidate := 3
        for candidate * candidate <= n {
            if n % candidate == 0 {
                return false
            }
            candidate += 2
        }
        return true
    }
}

// Uses a named return parameter.
func primes_less_than(n int) (result int) {
    for i := 2; i < n; i++ {
        if is_prime(i) {
            result++
        }
    }
    return // required with named return parameter
}

File: docs/languages/racket/lectures/primes.rb
# primes.rb

#
# Returns true if n is a prime number, false otherwise. Uses trial division to
# find factors.
#
# Ruby functions that return a boolean value customarily end with ?.
#

# def is_prime?(n)
#     if n < 2
#         false
#     elsif n == 2
#         return true
#     elsif n.even?
#         return false
#     else
#         candidate = 3
#         while candidate * candidate <= n
#             return false if n % candidate == 0
#             candidate += 2
#         end
#         return true
#     end
# end

#
# An alternative implementation using if as a modifier.
#
def is_prime?(n)
    return false if n < 2
    return true if n == 2
    return false if n.even?

    candidate = 3
    while candidate * candidate <= n
        return false if n % candidate == 0
        candidate += 2
    end
    return true
end


#
# Returns the number of primes less than the integer n.
#
# (1..n-1) is the range of integers from 1 up to, and including, n-1.
#
# (1..n-1).count {|i| is_prime?(i)} returns the number of ints from 1 to n-1
# that satisfy is_prime?, i.e. the number of primes from 1 to n-1 inclusive.
#
def primes_less_than(n)
    return (1..n-1).count {|i| is_prime?(i)}
end

#
# Which of the numbers from 0 to 30 are prime?
#
# Prints numbers and indicates which ones are prime.
#
count = 0
30.times do |n| 
    if is_prime?(n)
        count += 1
        puts "#{n} <--- prime #{count}"
    else
        puts "#{n}"
    end
end

max = 10000 # 1229 for 1000
puts "Number of primes less than #{max}: #{primes_less_than(max)}"
